src/v/kafka/server/quota_manager.cc

```diff
@@ -204,46 +202,62 @@ void quota_manager::update_client_quotas() {
 
 // If client is part of some group then client quota ID is a group
 // else client quota ID is client_id
-static std::optional<std::string_view> get_client_quota_id(
+static tracker_key get_client_quota_id(
   const std::optional<std::string_view>& client_id,
   const std::unordered_map<ss::sstring, config::client_group_quota>&
     group_quota) {
     if (!client_id) {
-        return std::nullopt;
+        // requests without a client id are grouped into an anonymous group that
+        // shares a default quota. the anonymous group is keyed on empty string.
+        return k_client_id{""};
```

Comment:
Ben can probably comment on this in more detail, but I seem to recall that returning different variant alternatives in this way can interfere with return value optimization.

If that's the case and I'm not making it up:

```suggestion
        return tracker_key{k_client_id{""}};
```

And similar elsewhere in this function.

Reply:
scratch that, two versions generate identical code.

Reply:
Ok, I have it: https://godbolt.org/z/f69MrhMG3

So this usage does break RVO, but so does my first suggestion. But this version should let RVO kick in:

```c++
        return tracker_key{std::in_place_type<k_client_id>, ""};
```

Reply:
Nice, thanks! I didn't know about `std::in_place_type`, it's pretty cool.

Reply:
Yeah, randomly gleaned from c++ weekly - https://youtu.be/0yJk5yfdih0?si=pEAUeTPOGRvG54GV

Kinda freaked me out when I saw it, honestly. I had been (and probably still am) in the habit of returning unwrapped `T{}` for `std::optional<T>`. 

It's interesting because `std::move`ing a local into a return statement I'm pretty sure has roughly the same effect, but we have a clang-tidy warning for that. Not sure whether a similar plugin exists to address this case or how tricky it would be to implement.

Reply:
> Yeah, randomly gleaned from c++ weekly 

Yeah, I saw that too, a bit frustrating.